### ISSUE TITLE
Handling docs >max tokens supported by a model for embeddings

### ISSUE THREAD

**Issue by zehawki** [typesense/typesense#1526]
I'm curious as to how TS supports generation of embeddings for documents with more tokens than the model can handle (most models support 512 only).

Will it just truncate the doc (or does the model do that automatically)?
If so, whats the best method to allow the longer documents?
OR, is this not a reasonable use case for TS, and its meant more for short text than long form?

Stories: https://stackoverflow.com/questions/58636587/how-to-use-bert-for-long-text-classification

**Comment by zehawki** [comment]
I can think of one way to solve for this is to perform chunking at source, ie split longer docs and send them in as shorter docs. Since this will prolly result in the same doc appearing in search results more than once, use a fake [group_by](https://typesense.org/docs/0.25.2/api/search.html#group-results) to dedup. But this is all very inelegant I suppose.

**Comment by ozanarmagan** [comment]
Since this is the most common solution to the problem, we [truncate](https://github.com/typesense/typesense/blob/261664fe110fb46506672420fc28fdd6e755ad69/src/text_embedder_tokenizer.cpp#L23) the document to 512 tokens after processing the BERT tokenizer.

>OR, is this not a reasonable use case for TS, and its meant more for short text than long form

This is actually a hard limit for BERT models to preserve the performance since self-attention mechanism grows quadratically against sequence length, so yes we can say that is meant more for shorter texts.

**Comment by zehawki** [comment]
Is there some new thought on this or some feature or hack that can be used? group_by isn't really working for chunking / dedup.

**Comment by kishorenc** [comment]
Even if a model supports longer than 512 token context-lengths, the embedding quality for large tokens of text will be poor because it's impossible to semantically represent many words in a small number of fixed dimensions. Chunking is how everyone is handling this at the moment.

**Comment by zehawki** [comment]
Yup, I'm also on board with chunking. What I meant above was - is there some feature or hack I can use for deduplicating results with chunking?

**Comment by kishorenc** [comment]
Can you please describe the current issues you are facing with `group_by` and chunking?

**Comment by zehawki** [comment]
Since this is a fake `group_by` based on some facet field that is artificially assigned, **there will be as many values as there are documents and potentially infinite**. Further, each time a doc is added, a never-seen-before facet value will be given to TS.

1. I don't know what impact, if any, this has on TS internals. Perhaps it was not designed for such a massive number of group_key / facets for a single field. One would imagine that TS architecture assumes, for efficiency, that number of facets per field are finite and limited and repetitive. 

And re the nested structure (https://typesense.org/docs/26.0/api/search.html#group-results): 

1. The `grouped_hits` returns an inflated json structure that is inelegant in this context - but OK one can live with it.
2. The main issue is that the number of results **count the chunked documents as well**, so if `per_page` is set as 10, one may actually get back any number of uniques between 1 and 10, depending on where each doc has been chunked. Here's a real example, which has only 4 unique docs. So it leads to poor UX since a different count of results will be displayed each time, not to mention headache in pagination.

![image](https://github.com/user-attachments/assets/52e399e3-e3c9-4121-812f-e102203f7c60)

To be sure, none of this is an issue with the actual feature of grouping - just force fitting to chunking.

**Comment by kishorenc** [comment]
> The main issue is that the number of results count the chunked documents as well, so if per_page is set as 10, one may actually get back any number of uniques between 1 and 10

Ok, I get what you mean. Since the documents are indexed at a per-chunk level in the vector index, when we query for top-K vectors (where `k` is set to `page * per_page` by default), we fetch the chunk-level vectors, which when mapped back to the group_by dimension, will result in fewer than `k` groups. 

>  So it leads to poor UX since a different count of results will be displayed each time, not to mention headache in pagination.

With vector search, pagination should be avoided, even when no chunking is involved. The nearest neighbor search on the HNSW vector index is an approximate search on a graph. When you increase the search radius during pagination (the `k` value), we could actually end up retrieving _better_ results that were not previously accessible with the smaller `k` value on the earlier page. This can actually cause the pagination to return duplicate results (since order changes).
